.github/workflows/build_and_push_docker_images.yml

```diff
@@ -17,26 +11,97 @@ concurrency:
 
 jobs:
   docker_build:
-    name: Build & publish Management
+    name: Build images
     runs-on: ubuntu-latest
 
     steps:
       - name: Checkout
         uses: actions/checkout@v2
 
-      - name: Login to container registry
-        uses: docker/login-action@v1
+      - name: Filter changes
+        uses: dorny/paths-filter@v2
+        id: filter
         with:
-          registry: ${{ secrets.docker_registry_server }}
-          username: ${{ secrets.docker_registry_username }}
-          password: ${{ secrets.docker_registry_password }}
+          filters: |
+            api:
+              - 'api_app/**/*'
+            resource_processor:
+              - 'resource_processor/**/*'
+            guacamole_server:
+              - 'templates/workspace_services/guacamole/guacamole-server/**/*'
+            gitea:
+              - 'templates/shared_services/gitea/**/*'
 
-      - name: Build & publish image
-        env:
-          DOCKER_BUILDKIT: 1
-          REPOSITORY_NAME: ${{ secrets.docker_registry_server }}/microsoft/azuretre/api
-        run: |
-          TAG=$GITHUB_SHA
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v1
 
-          docker build -t ${{ env.REPOSITORY_NAME}}:$TAG api_app/
-          docker push ${{ env.REPOSITORY_NAME}}:$TAG
+      - name: "Test image: api"
+        # test should be before build since its docker target is prior to runtime
+        if: (steps.filter.outputs.api == 'true' || github.event_name == 'workflow_dispatch')
+        uses: docker/build-push-action@v2
+        with:
+          context: ./api_app/
+          file: ./api_app/Dockerfile
+          outputs: type=local,dest=test-results
+          target: test-results
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+      - name: "Build image: api"
+        if: (steps.filter.outputs.api == 'true' || github.event_name == 'workflow_dispatch')
+        uses: docker/build-push-action@v2
+        with:
+          context: ./api_app/
+          file: ./api_app/Dockerfile
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+      - name: "Build image: resource_processor"
+        if: (steps.filter.outputs.resource_processor == 'true' || github.event_name == 'workflow_dispatch')
+        uses: docker/build-push-action@v2
+        with:
+          context: ./resource_processor
+          file: ./resource_processor/vmss_porter/Dockerfile
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+
+        # TODO: Add testing (#817) before this
```

Comment:
If we don't have tests here - does this mean that this PR goes untested? Why were they removed?

Reply:
There're tests but they aren't being triggered as part of the build. Anat is tasked to solve it.

Reply:
ahh, some other tests, was thinking this was our regular pytests that were removed